src/pm-conversations/pmConversationsSelectors.js

```diff
@@ -51,10 +67,114 @@ export const getRecentConversations: Selector<PmConversationData[]> = createSele
          end up converted to strings, so this access with string keys works.  We should probably use
          a Map for this and similar maps. */
         unreadPms[recipient.ids] || unreadHuddles[recipient.ids],
+    })),
+);
+
+/**
+ * Legacy implementation of {@link getRecentConversations}. Computes an
+ * approximation to the set of recent conversations, based on the messages we
+ * already know about.
+ */
+const getRecentConversationsLegacyImpl: Selector<PmConversationData[]> = createSelector(
+  getOwnUser,
+  getPrivateMessages,
+  getAttachUnread,
+  (ownUser: User, messages: Message[], attachUnread: UnreadAttacher): PmConversationData[] => {
+    const items = messages.map(msg => ({
+      ids: pmUnreadsKeyFromMessage(msg, ownUser.user_id),
+      recipients: normalizeRecipientsSansMe(msg.display_recipient, ownUser.email),
+      msgId: msg.id,
     }));
+
+    const sortedByMostRecent = collateByRecipient(items);
+
+    return attachUnread(sortedByMostRecent);
   },
 );
 
+/**
+ * Modern implementation of {@link getRecentConversations}. Returns exactly the
+ * most recent conversations. Requires server-side support.
+ */
+const getRecentConversationsImpl: Selector<PmConversationData[]> = createSelector(
+  getOwnUser,
+  getAllUsersById,
+  getRecentPrivateConversations,
+  getAttachUnread,
+  (
+    ownUser: User,
+    allUsers: Map<number, UserOrBot>,
+    recentPCs: RecentPrivateConversation[],
+    attachUnread: UnreadAttacher,
+  ) => {
+    const getEmail = (id: number): string => {
+      const user = allUsers.get(id);
+      if (user) {
+        return user.email;
+      }
+      throw new Error('getRecentConversations: unknown user id');
+    };
+
+    const recipients = recentPCs.map(conversation => {
+      const userIds = conversation.user_ids.slice();
+      if (userIds.length !== 1) {
+        userIds.push(ownUser.user_id);
+        userIds.sort((a, b) => a - b);
+      }
+
+      return {
+        ids: userIds.join(','),
+        recipients: normalizeRecipientsSansMe(
+          userIds.map(id => ({ email: getEmail(id) })),
+          ownUser.email,
+        ),
+        msgId: conversation.max_message_id,
+      };
+    });
+
+    return attachUnread(recipients);
+  },
+);
+
+/**
+ * The server version in which 'recent_private_conversations' was first made
+ * available.
+ */
+const DIVIDING_LINE = new ZulipVersion('2.1-dev-384-g4c3c669b41');
+
+// Private. Selector to choose between other selectors. (This avoids needlessly
+// recomputing the old version when we're on a new server, or vice versa.)
+const getMetaselector: Selector<Selector<PmConversationData[]>> = createSelector(
```

Comment:
Having this "metaselector" makes the code rather more complex than if we simply put this logic in `getRecentConversations`.

The thing it accomplishes is basically that the result of `isAtLeast` gets memoized. I'm dubious that that's much of a win -- we've designed that to precompute most of the work when constructing the `ZulipVersion` instance.

If you're keen on memoizing that, though, a straightforward way would be to do so directly:
```js
const getServerIsOld: Selector<boolean> = createSelector(
  getServerVersion,
  version => !(version && version.isAtLeast(DIVIDING_LINE),
);
```
(Or flip it around for a different name.) Then the memoized data is plain old data, and the control flow has fewer indirections to follow.